ruby/version.go

```diff
@@ -0,0 +1,170 @@
+package ruby
+
+import (
+	"errors"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+var (
+	anchoredVersion = regexp.MustCompile(`^\s*([0-9]+(\.[0-9a-zA-Z]+)*(-[0-9A-Za-z-]+(\.[0-9A-Za-z-]+)*)?)?\s*$`)
+
+	errInvalidVersion = errors.New("invalid gem version")
+)
+
+// Version is a RubyGem version.
+// This is based on the official implementation:
+//
+// https://github.com/rubygems/rubygems/blob/1a1948424aca90013b37cb8a78f5d1d5576023f1/lib/rubygems/version.rb
+type Version struct {
+	segs []Segment
+}
+
+func NewVersion(version string) (Version, error) {
+	if !anchoredVersion.MatchString(version) {
+		return Version{}, errInvalidVersion
+	}
+
+	version = strings.TrimSpace(version)
+	if version == "" {
+		version = "0"
+	}
+	version = strings.ReplaceAll(version, "-", ".pre.")
+
+	segs := canonicalize(version)
+	return Version{
+		segs: segs,
+	}, nil
+}
+
+func (v Version) Compare(other Version) int {
+	segs := v.segs
+	otherSegs := other.segs
+
+	leftLen := len(segs)
+	rightLen := len(otherSegs)
+	limit := max(leftLen, rightLen)
+	for i := 0; i < limit; i++ {
+		left, right := Segment(numericSegment(0)), Segment(numericSegment(0))
+		if i < leftLen {
+			left = segs[i]
+		}
+		if i < rightLen {
+			right = otherSegs[i]
+		}
+
+		if cmp := left.Compare(right); cmp != 0 {
+			return cmp
+		}
+	}
+
+	return 0
+}
+
+func canonicalize(v string) []Segment {
+	segs, prerelease := partitionSegments(v)
+
+	// Remove trailing zero segments.
+	i := len(segs) - 1
+	for ; i >= 0; i-- {
+		seg, ok := segs[i].(numericSegment)
+		if !ok || seg != 0 {
+			break
+		}
+	}
+	segs = segs[:i+1]
+
+	// Remove all zero segments preceding the first letter in a prerelease version.
+	if prerelease {
+		// Find the first letter in the version.
+		end := -1
+		for i := 0; i < len(segs); i++ {
+			if _, ok := segs[i].(stringSegment); ok {
+				end = i
+				break
+			}
+		}
+		if end != -1 {
+			// Find where the preceding zeroes start.
+			var start int
+			for i := end - 1; i >= 0; i-- {
+				seg, ok := segs[i].(numericSegment)
+				if !ok || seg != 0 {
+					start = i + 1
+					break
+				}
+			}
+			segs = append(segs[:start], segs[end:]...)
+		}
+	}
+
+	return segs
+}
+
+func partitionSegments(v string) (segs []Segment, prerelease bool) {
+	for _, s := range strings.Split(v, ".") {
+		if s == "" {
+			continue
+		}
+		n, err := strconv.ParseUint(s, 10, 64)
+		if err != nil {
```

Comment:
I'm not sure how we'd handle `ErrRange`, unfortunately. Just not support it?

Reply:
Another option would be to keep everything in string form and regex for digits and letters to determine how to proceed. It'd be safest, but probably less efficient since it deals with regexes

Reply:
It's annoying, but it seems like Ruby allows any arbitrarily long sequence of numbers be an `Integer`. Doesn't seem like there is a concept of `max_int`, so I guess it's possible this number is way larger than the largest 64 bit number.

I was considering keeping those arbitrarily long integers as strings and comparing from there, but I don't want to implement that comparison... Maybe we can just not support it?